streamx-flink/streamx-flink-connector/streamx-flink-connector-doris/src/main/java/com/streamxhub/streamx/flink/connector/doris/internal/DorisSinkFunction.java

```diff
@@ -19,167 +19,120 @@
+                return;
+            }
+            dorisSinkWriter.writeRecords(data.getDatabase(), data.getTable(), data.getDataRows());
         } else {
-            result = OBJECT_MAPPER.writeValueAsString(batch);
-        }
-
-        for (int i = 0; i <= maxRetries; i++) {
-            try {
-                dorisStreamLoad.load(result);
-                batch.clear();
-                break;
-            } catch (Exception e) {
-                LOGGER.error("doris sink error, retry times = {}", i, e);
-                if (i >= maxRetries) {
-                    throw new IOException(e);
-                }
-                try {
-                    Thread.sleep(1000 * i);
-                } catch (InterruptedException ex) {
-                    Thread.currentThread().interrupt();
-                    throw new IOException("unable to flush; interrupted while doing another attempt", e);
-                }
+            if (Strings.isNullOrEmpty(dorisConfig.database()) || Strings.isNullOrEmpty(dorisConfig.table())) {
+                throw new RuntimeException(" database|table  is empt ,please check your config or create DorisSinkRowDataWithMeta instance");
             }
+            dorisSinkWriter.writeRecords(dorisConfig.database(), dorisConfig.table(), (String) value);
         }
-    }
-
-    private void checkFlushException() {
-        if (flushException != null) {
-            throw new RuntimeException("Writing records to streamload failed.", flushException);
-        }
+        // raw data sink
+        totalInvokeRows.inc(1);
+        totalInvokeRowsTime.inc(System.nanoTime() - start);
+        return;
     }
 
     @Override
     public void close() throws Exception {
         super.close();
+        dorisSinkWriter.close();
+
     }
 
     @Override
-    public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throws Exception {
-        flush();
+    public void snapshotState(FunctionSnapshotContext context) throws Exception {
+        if (Semantic.EXACTLY_ONCE.equals(Semantic.of(dorisConfig.semantic()))) {
+            // save state
+            checkpointedState.add(dorisSinkWriter.getBufferedBatchMap());
+            flushPreviousState();
+            return;
+        }
     }
 
     @Override
-    public void initializeState(FunctionInitializationContext functionInitializationContext) throws Exception {
+    public void initializeState(FunctionInitializationContext context) throws Exception {
+        if (Semantic.EXACTLY_ONCE.equals(Semantic.of(dorisConfig.semantic()))) {
+            ListStateDescriptor<Map<String, DorisSinkBufferEntry>> descriptor =
+                new ListStateDescriptor<>(
+                    "buffered-rows",
+                    TypeInformation.of(new TypeHint<Map<String, DorisSinkBufferEntry>>() {
+                    })
+                );
```

Comment:
thanks for your pr,
Store data in state. When there is a lot of data and the checkpoint time is long, the state may be particularly large?
I think it's better to implement exactly-once based on [flink](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html) and [doris's](https://doris.apache.org/zh-CN/docs/data-operate/import/import-way/stream-load-manual.html) two-phase commit, what do you think?

Reply:
> thanks for your pr, Store data in state. When there is a lot of data and the checkpoint time is long, the state may be particularly large? I think it's better to implement exactly-once based on [flink](https://flink.apache.org/features/2018/03/01/end-to-end-exactly-once-apache-flink.html) and [doris's](https://doris.apache.org/zh-CN/docs/data-operate/import/import-way/stream-load-manual.html) two-phase commit, what do you think?

+ 1